anitya/tests/test_flask_api_v2.py

```diff
@@ -0,0 +1,254 @@
+            }
+        ]
+
+        self.assertEqual(data, exp)
+
+class AuthenticationRequiredTests(_APItestsMixin, Modeltests):
+    """Test anonymous access is blocked to APIs requiring authentication"""
+
+    def _check_authentication_failure_response(self, output):
+        data = _read_json(output)
+        # Temporary workaround for rendering-to-str on the server
+        data = json.loads(data)
+        # Check we get the expected error details
+        if self.oidc is None:
+            exp = {
+                'error': 'oidc_not_configured',
+                'error_description': 'OpenID Connect is not configured on the server'
+            }
+        else:
+            exp = {
+                "error_description": "Token required but invalid",
+                "error": "invalid_token",
+            }
+        self.assertEqual(data, exp)
+
+
+    def test_project_monitoring_request(self):
+        self.maxDiff = None
+        output = self.app.post('/api/v2/projects/')
+        self.assertEqual(output.status_code, 401)
+        self._check_authentication_failure_response(output)
+
+
+class _AuthenticatedAPItestsMixin(_APItestsMixin):
+    """Common test definitions for tests of authenticated access"""
+
+    def _post_app_url(self, post_url, **kwds):
+        return self.app.post(post_url, **kwds)
+
+    def test_invalid_project_monitoring_request(self):
+        # Check we get a 400 error reporting what we did wrong
+        # rather than an authentication error
+        output = self._post_app_url('/api/v2/projects/')
+        self.assertEqual(output.status_code, 400)
+        # Error details should report the missing required fields
+        data = _read_json(output)
+        error_details = data["message"]
+        self.assertIn("backend", error_details)
+        self.assertIn("homepage", error_details)
+        self.assertIn("name", error_details)
+
+    def test_valid_project_monitoring_request(self):
+        request_data = {
+            "backend": "PyPI",
+            "homepage": "http://python-requests.org",
+            "name": "requests",
+        }
+        output = self._post_app_url('/api/v2/projects/', data=request_data)
+        # Error details should report the missing required fields
+        data = _read_json(output)
+        print(data)
+        self.assertEqual(output.status_code, 200)
+
+
+class MockAuthenticationTests(_AuthenticatedAPItestsMixin, Modeltests):
+    """Test authenticated behaviour with authentication checks mocked out"""
+    def setUp(self):
+        super(MockAuthenticationTests, self).setUp()
+        # Replace anitya.authentication._validate_api_token
+        def _bypass_token_validation(validated, raw_api, *args, **kwds):
+            return raw_api(*args, **kwds)
+        mock_auth = mock.patch('anitya.authentication._validate_api_token',
+                               _bypass_token_validation)
+        mock_auth.start()
+        self.addCleanup(mock_auth.stop)
```

Comment:
Getting access to `addCleanup` regardless of Python version is one of the big reasons I always use the `unittest2` backport rather than the default `unittest` :)